### DISCUSSION TITLE
Remove the error message driven client behavior

### DISCUSSION THREAD

**Discussion by ahrtr** [etcd-io/etcd#14992] [General]
## Background

In order to fix https://github.com/etcd-io/etcd/issues/12385, PR https://github.com/etcd-io/etcd/pull/14322 introduced a change in which the client side may retry based on the error message returned from server side. 

This is not good, as it's too fragile and it's also changed the protocol between client and server. Please see the discussion in https://github.com/kubernetes/kubernetes/pull/114403

Note: The issue https://github.com/etcd-io/etcd/issues/12385 only happens when auth is enabled, and client side reuse the same client to watch.

### Solution 1
Rollback the change. Reasons:
1. K8s doesn't enable auth at all. It has no any impact on K8s.
2. It's very easy for client application to workaround the issue. The client just needs to create a new client each time before watching.

### Solution 2
Add a field something like `ReasonCode` in `WatchResponse`. https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/etcdserverpb/rpc.pb.go#L2192-L2221. The client side uses the ReasonCode instead of a fragile error message. 

### Thoughts ?
Please let me know your thought.

cc @mitake @ptabor @serathius @spzala  
@liggitt @dims @lavalamp @deads2k

**Comment by serathius** [comment]
Definitely first to do is rollback. This change should have never been backported to v3.5. We can continue the discussion about changing Watch api in v3.6.

**Reply by dims** [reply]
@serathius +1 to that plan of action, a fresh release on v3.5.x branch would be welcome as well.

**Comment by ahrtr** [comment]
FYI. https://github.com/etcd-io/etcd/pull/14995

**Comment by mitake** [comment]
So sorry for the trouble. I opened an issue for adding Code based member to WatchResponse: https://github.com/etcd-io/etcd/issues/15058

**Comment by lavacat** [comment]
@ahrtr We use etcd with auth for non-kube use-case. I got feedback that recreating client will be cumbersome. So, I'd like help contribute fix that will keep retry in the watch. Adding `ReasonCode` and not using Strings makes sense, but maybe retry logic can be added back with `ReasonCode`.

For wider community here are arguments why it will be useful to keep retry logic:
1. retry in the watch is consistent with etcd client that will retry in [interceptor](https://github.com/etcd-io/etcd/blob/6f899a7b40f7631461ffeda0067aa4c42dd17812/client/v3/retry_interceptor.go#L80).
2. [ErrAuthOldRevision](https://github.com/etcd-io/etcd/blob/26c06277932b77b1f0450c825af51229ba68d56e/server/etcdserver/v3_server.go#L653) requires user to understand etcd auth implementation details. 
3. If kube starts using auth, I think it will require some non-trivial refactoring to recreate the client. apiserver initializes etcd client in [factory](https://github.com/kubernetes/apiserver/blob/f928c72a9f84c5422dc1ea70919ded7570a99d12/pkg/storage/storagebackend/factory/etcd3.go#L297) package. But watch error handing is in [etcd3](https://github.com/kubernetes/apiserver/blob/f928c72a9f84c5422dc1ea70919ded7570a99d12/pkg/storage/etcd3/watcher.go#L175) package.

**Comment by ahrtr** [comment]
Thanks @lavacat , the direction of adding `ReasonCode` looks good to me. 

Please feel free to work on https://github.com/etcd-io/etcd/issues/15058 if @mitake hasn't get started.

**Reply by ahrtr** [reply]
Note that even K8s doesn't enable auth, it might still have some impact/concern on K8s because it changes the protocol between client and server. FYI. https://github.com/kubernetes/kubernetes/pull/114403#discussion_r1047194537

Let's follow the discussion in https://github.com/etcd-io/etcd/issues/15058
